src/stream/src/executor/source/kafka_backfill_executor.rs

```diff
@@ -0,0 +1,888 @@
+use std::time::Instant;
+
+use anyhow::anyhow;
+use either::Either;
+use futures::stream::{select_with_strategy, PollNext};
+use futures::StreamExt;
+use futures_async_stream::try_stream;
+use kafka_backfill_executor::source_executor::WAIT_BARRIER_MULTIPLE_TIMES;
+use risingwave_common::buffer::BitmapBuilder;
+use risingwave_common::metrics::GLOBAL_ERROR_METRICS;
+use risingwave_common::row::Row;
+use risingwave_common::system_param::local_manager::SystemParamsReaderRef;
+use risingwave_common::system_param::reader::SystemParamsRead;
+use risingwave_common::types::JsonbVal;
+use risingwave_connector::source::reader::desc::{SourceDesc, SourceDescBuilder};
+use risingwave_connector::source::{
+    BoxChunkSourceStream, SourceContext, SourceCtrlOpts, SplitId, SplitMetaData,
+};
+use risingwave_connector::ConnectorParams;
+use risingwave_storage::StateStore;
+use serde::{Deserialize, Serialize};
+use thiserror_ext::AsReport;
+
+use super::executor_core::StreamSourceCore;
+use super::kafka_backfill_state_table::BackfillStateTableHandler;
+use crate::executor::monitor::StreamingMetrics;
+use crate::executor::*;
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub enum BackfillState {
+    /// `None` means not started yet. It's the initial state.
+    Backfilling(Option<String>),
+    /// Backfill is stopped at this offset. Source needs to filter out messages before this offset.
+    SourceCachingUp(String),
+    Finished,
+}
+pub type BackfillStates = HashMap<SplitId, BackfillState>;
+
+impl BackfillState {
+    pub fn encode_to_json(self) -> JsonbVal {
+        serde_json::to_value(self).unwrap().into()
+    }
+
+    pub fn restore_from_json(value: JsonbVal) -> anyhow::Result<Self> {
+        serde_json::from_value(value.take()).map_err(|e| anyhow!(e))
+    }
+
+    /// Returns whether the row from upstream `SourceExecutor` is visible.
+    fn handle_upstream_row(&mut self, offset: &str) -> bool {
+        let mut vis = false;
+        match self {
+            BackfillState::Backfilling(None) => {
+                // backfilling for this split is not started yet. Ignore this row
+            }
+            BackfillState::Backfilling(Some(backfill_offset)) => {
+                match compare_kafka_offset(backfill_offset, offset) {
+                    Ordering::Less => {
+                        // continue backfilling. Ignore this row
+                    }
+                    Ordering::Equal => {
+                        // backfilling for this split is finished just right.
+                        *self = BackfillState::Finished;
+                    }
+                    Ordering::Greater => {
+                        // backfilling for this split produced more data than current source's progress.
+                        // We should stop backfilling, and filter out rows from upstream with offset <= backfill_offset.
+                        *self = BackfillState::SourceCachingUp(backfill_offset.clone());
+                    }
```

Comment:
Interesting, Could you please give some examples to show in which circumstances, this branch hits?

Reply:
e.g.,
1. Source Chunk: [1,2,3]
2. Backfil Chunk: [1,2,3,4,5]
3. Source Chunk: [4,5,6,7]